metricflow/query/group_by_item/candidate_push_down/group_by_item_candidate.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import annotations
+
+import itertools
+from dataclasses import dataclass
+from typing import Iterable, Tuple
+
+from typing_extensions import override
+
+from metricflow.query.group_by_item.path_prefixable import PathPrefixable
+from metricflow.query.group_by_item.resolution_dag.resolution_nodes.base_node import GroupByItemResolutionNode
+from metricflow.query.group_by_item.resolution_path import MetricFlowQueryResolutionPath
+from metricflow.specs.patterns.spec_pattern import SpecPattern
+from metricflow.specs.specs import InstanceSpecSet, LinkableInstanceSpec, LinkableSpecSet
+
+
+@dataclass(frozen=True)
+class GroupByItemCandidateSet(PathPrefixable):
+    """The set of candidate specs that could match a given spec pattern.
+
+    This candidate set is refined as it is passed from the root node (representing a measure) to the leaf node
+    (representing a query) in the resolution DAG.
+    """
+
+    specs: Tuple[LinkableInstanceSpec, ...]
+    # The path from the leaf node to the measure node where the specs originated. There can be multiple paths as
+    # when candidate sets are merged, there can be multiple measures were a spec could have come from.
+    measure_paths: Tuple[MetricFlowQueryResolutionPath, ...]
+    # The path from the leaf node to the node where this set was generated.
+    path_from_leaf_node: MetricFlowQueryResolutionPath
+
+    def __post_init__(self) -> None:  # noqa: D
+        # If there are no specs, there shouldn't be any measure paths.
+        assert (len(self.specs) > 0 and len(self.measure_paths) > 0) or (
+            len(self.specs) == 0 and len(self.measure_paths) == 0
+        )
+
+    @staticmethod
+    def intersection(
+        path_from_leaf_node: MetricFlowQueryResolutionPath, candidate_sets: Iterable[GroupByItemCandidateSet]
+    ) -> GroupByItemCandidateSet:
+        """Create a new candidate set that is the intersection of the given candidate sets.
+
+        The intersection is defined as the specs common to all candidate sets. path_from_leaf_node is used to indicate
+        where the new candidate set was created.
+        """
+        specs_as_sets = tuple(set(candidate_set.specs) for candidate_set in candidate_sets)
+        common_specs = set.intersection(*specs_as_sets)
+        if len(common_specs) == 0:
+            return GroupByItemCandidateSet.empty_instance()
+
+        measure_paths = tuple(
+            itertools.chain.from_iterable(candidate_set.measure_paths for candidate_set in candidate_sets)
+        )
+
+        return GroupByItemCandidateSet(
+            specs=tuple(common_specs), measure_paths=measure_paths, path_from_leaf_node=path_from_leaf_node
```

Comment:
So common_specs removes specs from the candidate input through the intersection, so there may be measure_paths that are no longer relevant to the output set. Just to make sure - we want to track the global measure_paths in all cases, correct?

Also I'm assuming order doesn't matter - like we want the measure_paths to preserve the chained order regardless of how the intersection re-orders the common_specs.